### PR TITLE
WPC-608: As an user i want to comment without authentication

### PR DESCRIPTION
**Pull request by elias-summermatter**

### CODE REVIEW

File: packages/editor/src/client/routes/commentList.tsx [hunk 1]
```diff
@@ -183,6 +183,9 @@ export function CommentList() {
     })
   }
 
+  let printUsername = ` ${currentComment?.anonymousName} (anonymous)`
+  if (currentComment?.user) printUsername = `${currentComment?.user.name}`
```
Comment (camilleory) [inline packages/editor/src/client/routes/commentList.tsx:187]: Translation for `anonymous`

File: packages/api/src/error.ts [hunk 2]
```diff
@@ -121,6 +122,24 @@ export class CommentLengthError extends ApolloError {
...
+
+export class AnonymousCommentsDisabledError extends ApolloError {
+  constructor() {
+    super(`Anonymous comments are disabled!`)
+  }
+}
+
+export class AnonymousCommentError extends ApolloError {
+  constructor() {
+    super(`You need to give an anonymousname if your not authenticated`)
```
Comment (penk85) [inline packages/api/src/error.ts:139]: typo here: "anonymous name", "you're".

File: packages/api/src/graphql/mutation.public.ts [hunk 3]
```diff
@@ -144,18 +147,37 @@ export const GraphQLPublicMutation = new GraphQLObjectType<undefined, Context>({
...
+            throw new AnonymousCommentsDisabledError()
+
+          if (!input.anonymousName) throw new AnonymousCommentError()
+          const captchaResult = await challenge.validateChallenge({
+            challengeID: input.challengeID,
+            solve: input.challengeSolution
+          })
+          if (!captchaResult.valid) throw new CommentAuthenticationError(captchaResult.msg)
+        } else {
+          input.anonymousName = ''
```
Comment (penk85) [inline packages/api/src/graphql/mutation.public.ts:170]: Can you explain why `anonymousName` gets cleared here?

File: packages/editor/src/client/panel/memberPlanEditPanel.tsx [hunk 4]
```diff
@@ -106,6 +106,7 @@ export function MemberPlanEditPanel({id, onClose, onSave}: MemberPlanEditPanelPr
 
   useEffect(() => {
     if (data?.memberPlan) {
+      if (!data.memberPlan.tags) data.memberPlan.tags = ['']
```
Comment (penk85) [inline packages/editor/src/client/panel/memberPlanEditPanel.tsx:109]: Why did you add this?

File: packages/editor/src/client/routes/commentList.tsx (hunk 1 above)
Comment (penk85) [reply packages/editor/src/client/routes/commentList.tsx:187]: Maybe better to display as 'unregistered user' if there's a display name?

Comment (penk85) [reply packages/editor/src/client/routes/commentList.tsx:187]: For simplicity would a ternary assignment work better here?

File: packages/editor/src/client/panel/memberPlanEditPanel.tsx (hunk 4 above)
Comment (elias-summermatter) [reply packages/editor/src/client/panel/memberPlanEditPanel.tsx]: My IDE does weird stuff... it changed something else which made this change required have fixed this now

File: packages/api/src/error.ts [hunk 5]
```diff
@@ -19,7 +19,8 @@ export enum ErrorCode {
   CommentLengthError = 'COMMENT_LENGTH_ERROR',
   PeerTokenInvalid = 'PEER_TOKEN_INVALID',
   InternalError = 'InternalError',
-  UserSubscriptionAlreadyDeactivated = 'USER_SUBSCRIPTION_ALREADY_DEACTIVATED'
+  UserSubscriptionAlreadyDeactivated = 'USER_SUBSCRIPTION_ALREADY_DEACTIVATED',
+  InvalidCaptcha = 'InvalidCaptcha'
```
Comment (tomaszdurka) [inline packages/api/src/error.ts:23]: Again "ChallengeFailed" ?

**Review by tomaszdurka — CHANGES_REQUESTED** [review]

File: packages/api/src/context.ts [hunk 6]
```diff
@@ -105,12 +106,14 @@ export interface Context {
...
   readonly paymentProviders: PaymentProvider[]
   readonly hooks?: Hooks
+  readonly challenge: ChallengeProvider
 
   getOauth2Clients(): Promise<OAuth2Clients[]>
 
   authenticate(): Session
   authenticateToken(): TokenSession
   authenticateUser(): UserSession
+  optionalAuthenticateUser(): UserSession | null
```
[the remark this answers is not in the document]
Comment (elias-summermatter) [reply packages/api/src/context.ts:116]: Why rename? Its an additional middleware.

File: packages/api/src/graphql/query.public.ts [hunk 7]
```diff
@@ -426,6 +427,22 @@ export const GraphQLPublicQuery = new GraphQLObjectType<undefined, Context>({
         const updatedInvoices = await dbAdapter.invoice.getInvoicesByUserID(user.id)
         return updatedInvoices.find(invoice => invoice !== null && invoice.id === id)
       }
+    },
+
+    // Challenge
```
[the remark this answers is not in the document]
Comment (elias-summermatter) [reply packages/api/src/graphql/query.public.ts:432]: You mean all comments or just "Challenge"?

File: packages/api/src/challenges/challengeProvider.ts [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+}
+export type ChallengeValidationProps = {
+  challengeID: string
+  solve: number
+}
+
+export type ChallengeValidationReturn = {
+  result: string
+  msg: string
+  valid: boolean
```
[the remark this answers is not in the document]
Comment (elias-summermatter) [reply packages/api/src/challenges/challengeProvider.ts:14]: msg or new message is human readable message while result is a string to react on a specific reason why it failed. We don't use this for now but it can be interesting in the future. I can remove this but i guess at some point we will be happy to have this.

**Review by michael-scheurer — DISMISSED** [review]

File: packages/api/src/graphql/comment.ts [hunk 9]
```diff
@@ -97,7 +97,9 @@ export const GraphQLPublicCommentInput = new GraphQLInputObjectType({
   name: 'CommentInput',
   fields: {
     parentID: {type: GraphQLID},
-
+    anonymousName: {type: GraphQLString},
+    challengeID: {type: GraphQLString},
```
[the remark this answers is not in the document]
Comment (elias-summermatter) [reply packages/api/src/graphql/comment.ts:101]: Did this

**Comment by elias-summermatter** [comment]
@tomaszdurka can you please specify what is needed that this pr is ready to merge?

**Review by tomaszdurka — CHANGES_REQUESTED** [review]
So let's wrap it up again:
1. As discussed before this doesn't hold with the `authorType: VerifiedUser` as it's not really a user. Add here `GuestUser` value.
2. Rename `anonymousName` to `username`, keep it empty for comments with user object as you suggested.

Going academic:
Actually I think this whole structure should be within field `author` which would be an object with field `authorType: VerifiedUser | GuestUser | Team | Author` which would then discriminate other properties like `user` or username`.
But that's breaking a lot and not worth hassle.

**Comment by elias-summermatter** [comment]
> So let's wrap it up again:
> 
>     1. As discussed before this doesn't hold with the `authorType: VerifiedUser` as it's not really a user. Add here `GuestUser` value.
> 
>     2. Rename `anonymousName` to `username`, keep it empty for comments with user object as you suggested.
> 
> 
> Going academic: Actually I think this whole structure should be within field `author` which would be an object with field `authorType: VerifiedUser | GuestUser | Team | Author` which would then discriminate other properties like `user` or username`. But that's breaking a lot and not worth hassle.

I changed it this way is that fine like this... I Named it guestUsername to  make it clear and get not mixed up with other usernames

**Review by tomaszdurka — APPROVED** [review]

**Review by michael-scheurer — APPROVED** [review]
